File: typescript/src/manipulate/manipulate.ts
export function manipulate(text: string): string[] {
  const result: string[] = [];
  const chars = [...text];

  function permutations(index: number, arr: string[]): void {
    if (index === arr.length - 1) {
      result.push(arr.join(''));
      return;
    }

    const seen = new Set<string>();

    for (let i = index; i < arr.length; i++) {
      if (seen.has(arr[i])) continue;
      seen.add(arr[i]);

      // swap
      [arr[index], arr[i]] = [arr[i], arr[index]];
      permutations(index + 1, arr);
      [arr[index], arr[i]] = [arr[i], arr[index]];
    }
  }

  permutations(0, chars);
  result.sort();
  return result;
}

File: golang/manipulate/manipulate.go
package manipulate

import "sort"

// https://www.geeksforgeeks.org/print-all-permutations-of-a-string-with-duplicates-allowed-in-input-string/
// https://www.geeksforgeeks.org/write-a-c-program-to-print-all-permutations-of-a-given-string/

func permutations(index int, sliceCharacters []rune, result *[]string) {
	if index == len(sliceCharacters)-1 {
		*result = append(*result, string(sliceCharacters))
		return
	}

	seen := make(map[rune]bool)

	for i := index; i < len(sliceCharacters); i++ {
		if seen[sliceCharacters[i]] {
			continue
		}

		// prevent duplicated
		seen[sliceCharacters[i]] = true

		// swap
		sliceCharacters[index], sliceCharacters[i] = sliceCharacters[i], sliceCharacters[index]
		permutations(index+1, sliceCharacters, result)
		sliceCharacters[index], sliceCharacters[i] = sliceCharacters[i], sliceCharacters[index]
	}

}

func Manipulate(text string) []string {
	result := []string{}
	permutations(0, []rune(text), &result)
	sort.Strings(result)

	return result
}
